drivers/serial/uart_nrf5e.c

```diff
@@ -0,0 +1,873 @@
+
+		while (!uart->EVENTS_ENDRX) {
+		}
+
+		uart->EVENTS_ENDRX = 0;
+	} while (uart->RXDAMOUNT == 0);
+
+	return 0;
+
+#else
+	if (uart_nrf5e_fifo_read(dev, c, 1)) {
+		return 0;
+	}
+	return -1;
+#endif
+}
+
+/**
+ * @brief Output a character in polled mode.
+ *
+ * Checks if the transmitter is empty. If empty, a character is written to
+ * the data register.
+ *
+ * @param dev UART device struct
+ * @param c Character to send
+ *
+ * @return Sent character
+ */
+
+static unsigned char uart_nrf5e_poll_out(struct device *dev,
+					unsigned char c)
+{
+#ifndef CONFIG_UART_INTERRUPT_DRIVEN
+	volatile struct _uart *uart = UART_STRUCT(dev);
+
+	do {
+		uart->TXD = (u32_t)&c;
+		uart->TXDMAXCNT = 1;
+		uart->TASKS_STARTTX = 1;
+
+		while (!uart->EVENTS_ENDTX) {
+		}
+
+		uart->EVENTS_ENDTX = 0;
+	} while (uart->TXDAMOUNT == 0);
+
+#else
+	/* If the queue is full we need to pump it to make room */
+	uart_nrf5e_mask(dev, 1);
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+
+	while (!uart_nrf5e_fifo_fill(dev, &c, 1)) {
+		if (dev_data->in_interrupt) {
+			uart_nrf5e_pump_tx_fifo(dev);
+		}
+	}
+
+#endif
+
+	return c;
+}
+
+/** Console I/O function */
+static int uart_nrf5e_err_check(struct device *dev)
+{
+	volatile struct _uart *uart = UART_STRUCT(dev);
+	u32_t error = 0;
+
+	if (uart->EVENTS_ERROR) {
+		/* register bitfields maps to the defines in uart.h */
+		error = uart->ERRORSRC;
+
+		/* Clear the register */
+		uart->ERRORSRC = error;
+	}
+
+	error = error & 0x0F;
+
+	return error;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+static void uart_nrf5e_fifo_tx_start(struct device *dev)
+{
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+	volatile struct _uart *uart = UART_STRUCT(dev);
+
+	if (dev_data->fifo_tx_operating) {
+		return;
+	}
+
+	if (dev_data->fifo_tx_read_index == dev_data->fifo_tx_write_index) {
+		return;
+	}
+
+	dev_data->fifo_tx_operating = 1;
+
+	uart->TXD =
+		(u32_t)&dev_data->fifo_tx_data[dev_data->fifo_tx_read_index];
+
+	u8_t tx_size = 0;
+
+	if (dev_data->fifo_tx_write_index < dev_data->fifo_tx_read_index) {
+		/* We can go till the end of the buffer */
+		tx_size = sizeof(dev_data->fifo_tx_data) -
+			  dev_data->fifo_tx_read_index;
+	} else {
+		/* We can read as many bytes as needed to
+		 * make read index equal to write index
+		 */
+		tx_size = dev_data->fifo_tx_write_index -
+			  dev_data->fifo_tx_read_index;
+	}
+
+	/* Limit the blocklength */
+	if (tx_size > CONFIG_UART_NRF5E_TX_DMA_MAXBLOCK) {
+		tx_size = CONFIG_UART_NRF5E_TX_DMA_MAXBLOCK;
+	}
+
+	/* And go */
+	uart->TXDMAXCNT = tx_size;
+	uart->TASKS_STARTTX = 1;
+}
+
+/** Interrupt driven FIFO fill function */
+static int uart_nrf5e_fifo_fill(struct device *dev,
+			 const u8_t *tx_data, int len)
+{
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+	u8_t num_tx = 0;
+
+	if (!tx_data) {
+		return 0;
+	}
+
+	uart_nrf5e_mask(dev, 1);
+
+	while (len - num_tx) {
+		u8_t new_index = dev_data->fifo_tx_write_index + 1;
```

Comment:
Since fifo_tx_write_index is always incremented by 1, I think on line 495 we could replace >= with ==